File: pyround1/783.二叉搜索树节点最小距离.py
#
# @lc app=leetcode.cn id=783 lang=python3
#
# [783] 二叉搜索树节点最小距离
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def minDiffInBST(self, root: TreeNode) -> int:
        vals = []
        def dfs(node):
            if node:
                vals.append(node.val)
                dfs(node.left)
                dfs(node.right)
        
        dfs(root)
        vals.sort()
        return min(vals[i+1] - vals[i] for i in range(len(vals) - 1))
# @lc code=end

File: 783.二叉搜索树节点最小距离.cpp
/*
 * @lc app=leetcode.cn id=783 lang=cpp
 *
 * [783] 二叉搜索树节点最小距离
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int prev;
    int res = INT_MAX;
    int minDiffInBST(TreeNode* root) {
        prev = 0;
        helper(root);
        return res;
    }
    void helper(TreeNode* root){
        if(!root)
            return ;
        if(root->left)
        {
            helper(root->left);
        }
        if(prev)
            res = min(res, root->val - prev);
        prev = root->val;
        if(root->right)
        {
            helper(root->right);
        }

    }
};
// @lc code=end

